cmd/src/batch_exec.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/sourcegraph/src-cli/internal/cmderrors"
+
+	batcheslib "github.com/sourcegraph/sourcegraph/lib/batches"
+)
+
+func init() {
+	usage := `
+INTERNAL USE ONLY: 'src batch exec' executes the given raw batch spec in the given workspaces.
+
+The input file contains a JSON dump of the WorkspacesExecutionInput struct in
+github.com/sourcegraph/sourcegraph/lib/batches.
+
+Usage:
+
+    src batch exec -f FILE [command options]
+
+Examples:
+
+    $ src batch exec -f batch-spec-with-workspaces.json
+
+`
+
+	flagSet := flag.NewFlagSet("exec", flag.ExitOnError)
+	flags := newBatchExecuteFlags(flagSet, true, batchDefaultCacheDir(), batchDefaultTempDirPrefix())
+
+	handler := func(args []string) error {
+		if err := flagSet.Parse(args); err != nil {
+			return err
+		}
+
+		if len(flagSet.Args()) != 0 {
+			return cmderrors.Usage("additional arguments not allowed")
+		}
+
+		ctx, cancel := contextCancelOnInterrupt(context.Background())
+		defer cancel()
+
+		err := executeBatchSpecInWorkspaces(ctx, executeBatchSpecOpts{
+			flags:  flags,
+			client: cfg.apiClient(flags.api, flagSet.Output()),
+			ui:     &ui.JSONLines{},
+		})
+		if err != nil {
+			return cmderrors.ExitCode(1, nil)
+		}
+
+		return nil
+	}
+
+	batchCommands = append(batchCommands, &command{
+		flagSet: flagSet,
+		handler: handler,
+		usageFunc: func() {
+			fmt.Fprintf(flag.CommandLine.Output(), "Usage of 'src batch %s':\n", flagSet.Name())
+			flagSet.PrintDefaults()
+			fmt.Println(usage)
+		},
+	})
+}
+
+func executeBatchSpecInWorkspaces(ctx context.Context, opts executeBatchSpecOpts) (err error) {
+	defer func() {
+		if err != nil {
+			opts.ui.ExecutionError(err)
+		}
+	}()
+
+	svc := service.New(&service.Opts{
+		AllowUnsupported: opts.flags.allowUnsupported,
+		AllowIgnored:     opts.flags.allowIgnored,
+		Client:           opts.client,
+	})
+	if err := svc.DetermineFeatureFlags(ctx); err != nil {
+		return err
+	}
+
+	if err := checkExecutable("git", "version"); err != nil {
+		return err
+	}
+	if err := checkExecutable("docker", "version"); err != nil {
+		return err
+	}
+
+	// Read the input file that contains the raw spec and the workspaces in
+	// which to execute it.
+	input, err := loadWorkspaceExecutionInput(opts.flags.file)
+	if err != nil {
+		return err
+	}
+
+	// Since we already know which workspaces we want to execute the steps in,
+	// we can convert them to RepoWorkspaces and build tasks only for those.
+	repoWorkspaces := convertWorkspaces(input.Workspaces)
+
+	// Parse the raw batch spec contained in the input
+	opts.ui.ParsingBatchSpec()
+	batchSpec, err := svc.ParseBatchSpec([]byte(input.RawSpec))
+	if err != nil {
+		if merr, ok := err.(*multierror.Error); ok {
+			opts.ui.ParsingBatchSpecFailure(merr)
+			return cmderrors.ExitCode(2, nil)
+		} else {
+			// This shouldn't happen; let's just punt and let the normal
+			// rendering occur.
+			return err
+		}
+	}
+	opts.ui.ParsingBatchSpecSuccess()
```

Comment:
idea: We could share these with batch_common, if we made these separate functions. Might also read nicer in the main loop and reduces duplication:

```
parsingBatchSpec(ui, batchSpec)
preparingContainerImages(ui, svc, batchSpec)
..
```

Reply:
Okay, we can try that, but let's do that in another PR, since that's tricky stuff with some of the error handling etc.